src/drove.go

```diff
@@ -133,175 +143,204 @@ func fetchRecentEvents(client *http.Client, syncPoint *CurrSyncPoint) (*DroveEve
 			},
+		},
+		eventRefreshInterval: config.EventRefreshIntervalSec,
+	}
+}
+
+func pollingHandler(droveClient *DroveClient) {
+	namespace := droveClient.namespace
+	logger.WithFields(logrus.Fields{
+		"at":        time.Now(),
+		"namespace": namespace,
+	}).Debug("Syncing...")
+
+	droveClient.syncPoint.Lock()
+	defer droveClient.syncPoint.Unlock()
+
+	leaderShifted := refreshLeaderData(namespace)
+	eventSummary, err := fetchRecentEvents(droveClient.httpClient, &droveClient.syncPoint, namespace)
+	if err != nil {
+		logger.WithFields(logrus.Fields{
+			"error":     err.Error(),
+			"namespace": namespace,
+		}).Error("unable to sync events from drove")
+	} else {
+		if len(eventSummary.EventsCount) > 0 {
+			logger.WithFields(logrus.Fields{
+				"events":    eventSummary.EventsCount,
+				"namespace": namespace,
+				"localTime": time.Now(),
+			}).Info("Events received")
+			reloadNeeded := false
+			if _, ok := eventSummary.EventsCount["APP_STATE_CHANGE"]; ok {
+				reloadNeeded = true
+			}
+			if _, ok := eventSummary.EventsCount["INSTANCE_STATE_CHANGE"]; ok {
+				reloadNeeded = true
+			}
+			if reloadNeeded || leaderShifted {
+				refreshApps(droveClient.httpClient, namespace, leaderShifted)
+			} else {
+				logger.Debug("Irrelevant events ignored")
+			}
+		} else {
+			logger.WithFields(logrus.Fields{
+				"events":    eventSummary.EventsCount,
+				"namespace": namespace,
+			}).Debug("New Events received")
 		}
-		syncData := CurrSyncPoint{}
-		refreshInterval := 2
-		if config.EventRefreshIntervalSec > 0 {
-			refreshInterval = config.EventRefreshIntervalSec
-		}
-		ticker := time.NewTicker(time.Duration(refreshInterval) * time.Second)
-		for range ticker.C {
-			func() {
-				logger.WithFields(logrus.Fields{
-					"at": time.Now(),
-				}).Debug("Syncing...")
-				syncData.Lock()
-				defer syncData.Unlock()
-				refreshLeaderData()
-				eventSummary, err := fetchRecentEvents(client, &syncData)
-				if err != nil {
-					logger.WithFields(logrus.Fields{
-						"error": err.Error(),
-					}).Error("unable to sync events from drove")
-				} else {
-					if len(eventSummary.EventsCount) > 0 {
-						logger.WithFields(logrus.Fields{
-							"events":    eventSummary.EventsCount,
-							"localTime": time.Now(),
-						}).Info("Events received")
-						reloadNeeded := false
-						if _, ok := eventSummary.EventsCount["APP_STATE_CHANGE"]; ok {
-							reloadNeeded = true
-						}
-						if _, ok := eventSummary.EventsCount["INSTANCE_STATE_CHANGE"]; ok {
-							reloadNeeded = true
-						}
-						if reloadNeeded {
-							select {
-							case eventqueue <- true: // add reload to our queue channel, unless it is full of course.
-							default:
-								logger.Warn("queue is full")
-							}
-						} else {
-							logger.Debug("Irrelevant events ignored")
-						}
-					} else {
-						logger.WithFields(logrus.Fields{
-							"events": eventSummary.EventsCount,
-						}).Debug("New Events received")
-					}
-				}
-			}()
-		}
-	}()
+	}
```

Comment:
The eventqueue is removed? what if reload takes more than 5 secs?

Reply:
added